src/service_discovery/mod.rs

```diff
@@ -95,45 +104,38 @@ impl ServiceDiscovery {
     pub fn seek_peers(&mut self) -> Result<(), ServiceDiscoveryError> {
         let _ = self
             .socket
-            .send_to(&self.seek_peers_req, &self.remote_addr)?;
+            .write_to(Some((&self.seek_peers_req, self.remote_addr, 0)))?;
         Ok(())
     }
 
     /// Register service discovery observer
-    pub fn register_observer(&mut self, obs: Sender<Vec<SocketAddr>>) {
+    pub fn register_observer(&mut self, obs: Sender<Vec<PeerInfo>>) {
         self.observers.push(obs);
     }
 
     fn read(&mut self, core: &mut Core, poll: &Poll) {
-        let (bytes_rxd, peer_addr) = match self.socket.recv_from(&mut self.read_buf) {
-            Ok((bytes_rxd, peer_addr)) => (bytes_rxd, peer_addr),
-            // TODO(povilas): handle WouldBlock as well
-            Err(ref e) if e.kind() == ErrorKind::Interrupted => return,
-            Err(e) => {
-                debug!("ServiceDiscovery error in read: {:?}", e);
-                self.terminate(core, poll);
-                return;
-            }
-        };
-
-        let msg: DiscoveryMsg = match deserialise(&self.read_buf[..bytes_rxd]) {
-            Ok(msg) => msg,
-            Err(e) => {
-                debug!("Bogus message serialisation error: {:?}", e);
-                return;
-            }
-        };
-
-        match msg {
-            DiscoveryMsg::Request { guid } => {
-                if self.listen && self.guid != guid {
-                    self.reply_to.push_back(peer_addr);
-                    self.write(core, poll)
+        loop {
+            let (msg, peer_addr) = match self.socket.read_frm() {
+                Ok(Some((msg, peer_addr))) => (msg, peer_addr),
+                Ok(None) => return,
+                Err(e) => {
+                    debug!("ServiceDiscovery error in read: {:?}", e);
+                    self.terminate(core, poll);
+                    return;
+                }
+            };
+
+            match msg {
+                DiscoveryMsg::Request { our_pk: their_pk } => {
+                    if self.listen && self.our_pk != their_pk {
+                        self.reply_to.push_back(peer_addr);
```

Comment:
Don't think you need to do all this book keeping if using `UdpSock` from sock-coll.

Reply:
meaning just write directly to `UdpSock`, because it will buffer writes anyway? :)

Reply:
It'll buffer it only if it can't write it. So `write()` here should take an option of message. This will pass `Some()` for that and write being called from the readiness handler will pass `None` meaning if it's a spurious wake then sock-coll will internally do nothing and if it was due to a prior WB then sock-coll would internally re-attempt to write the previously buffered stuff. [Check here](https://github.com/maidsafe/socket-collection/blob/e1ba943ded9b3e3b29bc14ba12b517aff9729b7b/src/udp.rs#L480).

So follow other modules [like here](https://github.com/maidsafe/crust/blob/a3d0deeeaf47739b482790e6074476af965b8162/src/main/active_connection.rs#L153) and a `None` case for it [here](https://github.com/maidsafe/crust/blob/a3d0deeeaf47739b482790e6074476af965b8162/src/main/active_connection.rs#L178)